File: Linked_List_free_code_camp/3.linked_list_sum.js
class Node {
    constructor(val, next = null) {
        this.val = val;
        this.next = next;
    }
}

const head =
    new Node(1,
        new Node(2,
            new Node(3,
                new Node(4, null))));

// A -> B -> C -> D -> null

const sumList = (head) => {
    let sum = 0;
    let curr = head;
    while (curr) {
        sum += curr.val;
        curr = curr.next;
    }
    return sum;
};

console.log(sumList(head));


const sumListRecursive = (head) => {
    if (head == null) return 0;
    const sum = sumListRecursive(head.next);
    return head.val + sum;
};

console.log(sumListRecursive(head));

File: Linked_List_free_code_camp/3.linked_list_sum.cpp
#include<bits/stdc++.h>
using namespace std;

class Node {
public:
    int val;
    Node* next;

    Node(int val, Node* next = NULL) {
        this->val = val;
        this->next = next;
    }
};

void print_v(vector<int> values) {
    for (auto e : values) {
        cout << e << " ";
    }
    cout << endl;
}

int linkedListSumIterative(Node* head) {
    int sum = 0;

    Node* curr = head;
    while (curr) {
        sum += curr->val;
        curr = curr->next;
    }
    return sum;
};

int linkedListSumRecursive(Node* head) {
    if (head == NULL) return 0;

    int faith = linkedListSumRecursive(head->next);

    return head->val + faith;
};


int main() {

    Node* head =
        new Node(10,
            new Node(20,
                new Node(30,
                    new Node(40))));

    cout << linkedListSumIterative(head) << endl;
    cout << linkedListSumRecursive(head) << endl;

    return 0;
}